Source/Foundation/SDPromise.m

```diff
@@ -23,6 +23,14 @@ - (void) resolve:(id)result;
 
 @end
 
+@interface SDPromiseFail : NSObject
+
+- (instancetype) initWithBlock:(SDPromiseFailBlock)block resultPromise:(SDPromise *)promise;
+
+- (void) reject:(NSError *)error;
+
+@end
```

Comment:
It's late for me right now so my judgment may be off, but looking at this it seems like SDPromiseThen and SDPromiseFail can be one class, and you'd only need one instance for each promise. It'd take both the Then and the Fail block and the resulting SDPromise. But the rest of the logic for these classes is either in resolve: or fail:. Maybe call it SDPromiseResult or something. You'd only need one NSMutableArray in SDPromise itself: instead of thenBlocks or failedBlocks, it could just be resultBlocks, which you'd then call the appropriate reject: or resolve: method on.